File: Level 1/Valid Parentheses.py
def is_valid(s):
    stack = []
    mp = {')':'(', '}':'{', ']':'['}

    for ch in s:
        if ch in mp.values():
            stack.append(ch)
        else:
            if not stack or stack[-1] != mp[ch]:
                return False
            stack.pop()

    return True if not stack else False


s = input()
print(is_valid(s))

File: Stack/20-valid-parentheses/20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> pairs = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        stack<char> st;

        for (int i = 0; i < s.length(); i++) {

            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                st.push(s[i]);
            }
            else {
                if (st.empty() || st.top() != pairs[s[i]]) {
                    return false;
                }

                st.pop();
            }
        }

        return st.empty();
    }   

};     